python/pyiceberg/io/pyarrow.py

```diff
@@ -532,131 +535,116 @@ def project_table(
+            return values.cast(schema_to_pyarrow(promote(file_field.field_type, field.field_type)))
 
-        Raises:
-            NotImplementedError: If attempting to visit an unrecognized object type
-        """
-        raise NotImplementedError(f"Cannot visit non-type: {requested_schema}")
+        return values
 
-    @visit_with_arrow.register(Schema)
-    def _(self, requested_schema: Schema, file_schema: Schema) -> pa.Table:
-        """Visit a Schema with a concrete SchemaVisitorWithPartner"""
-        struct_result = self.visit_with_arrow(requested_schema.as_struct(), file_schema.as_struct())
-        pyarrow_schema = schema_to_pyarrow(requested_schema)
-        return pa.Table.from_arrays(struct_result.flatten(), schema=pyarrow_schema)
+    def schema(self, schema: Schema, schema_partner: Optional[pa.Array], struct_result: Optional[pa.Array]) -> Optional[pa.Array]:
+        return struct_result
 
-    def _get_field_by_id(self, field_id: int) -> Optional[NestedField]:
-        try:
-            return self.file_schema.find_field(field_id)
-        except ValueError:
-            # Field is not in the file
+    def struct(
+        self, struct: StructType, struct_array: Optional[pa.Array], field_results: List[Optional[pa.Array]]
+    ) -> Optional[pa.Array]:
+        if struct_array is None:
             return None
 
-    @visit_with_arrow.register(StructType)
-    def _(self, requested_struct: StructType, file_struct: Optional[IcebergType]) -> pa.Array:  # pylint: disable=unused-argument
-        """Visit a StructType with a concrete SchemaVisitorWithPartner"""
-        results = []
-
-        for requested_field in requested_struct.fields:
-            file_field = self._get_field_by_id(requested_field.field_id)
-
-            if file_field is None and requested_field.required:
-                raise ResolveException(f"Field is required, and could not be found in the file: {requested_field}")
-
-            results.append(self.visit_with_arrow(requested_field.field_type, file_field))
-
-        pyarrow_schema = schema_to_pyarrow(requested_struct)
-        return pa.StructArray.from_arrays(arrays=results, fields=pyarrow_schema)
-
-    @visit_with_arrow.register(ListType)
-    def _(self, requested_list: ListType, file_field: Optional[NestedField]) -> pa.Array:
-        """Visit a ListType with a concrete SchemaVisitorWithPartner"""
+        field_arrays: List[pa.Array] = []
+        fields: List[pa.Field] = []
+        for pos, field_array in enumerate(field_results):
+            field = struct.fields[pos]
+            if field_array is not None:
+                array = self.cast_if_needed(field, field_array)
+                field_arrays.append(array)
+                fields.append(pa.field(field.name, array.type, field.optional))
+            elif field.optional:
+                arrow_type = schema_to_pyarrow(field.field_type)
+                field_arrays.append(pa.nulls(len(struct_array)).cast(arrow_type))
```

Comment:
Note that this differs. It uses the length of the current partner rather than the table's length. I think this is more correct because this could be nested and have a length other than the number of rows.

Reply:
Instead of casting, we can just set the type
```suggestion
                field_arrays.append(pa.nulls(len(struct_array), type=arrow_type))
```

For a PyArrow table, all the arrays need to be the same length.

> An example test case is when the schema has been updated and has added a field to a struct inside a list. For example, locations list<lat: float, long: float>. If there are 3 rows, but each row has 2 locations, then there should be 6 values for each lat and long. Then if an "alt" field is added, it also needs 6 values.

Wouldn't alt be just yet another array?

Reply:
Yes, `alt` would be another array of nulls that we add. But the difference is that we need `alt` to be 6 values, not 3 values because it has 2 values per table row.